File: Public/Gui/Linn/Gui/Resources/TextureArray.py
import Plugin
import TextureManager

class TextureArray(Plugin.Plugin):
    def __init__(self, aName=None, aTextureList=None, aFps=None, aLoop=None):
        if aName == None:
            aName = self.UniqueName('TextureArray')
        Plugin.Plugin.__init__(self, aName)
        if aTextureList:
            self.iTextureList = aTextureList
        else:
            self.iTextureList = []
        self.iNumTextures = len(self.iTextureList)
        self.iTextureIndex = 0
    
    def Load(self, aStream):
        Plugin.Plugin.Load(self, aStream)
        
        textures = aStream.LoadNode('Texture')
        for texture in textures:
            self.iTextureList.append(texture.firstChild.nodeValue)
    
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'TextureArray')
            
        Plugin.Plugin.Save(self, aStream)
        
        for texture in self.iTextureList:
            textureRef = aStream.CreateElement('Texture')
            textureRef.appendChild(aStream.CreateTextNode(texture.Filename()))
            aStream.SaveNode(textureRef)
        
    def Link(self):
        #print 'Linking', self.Fullname(), '[TextureArray]'
        Plugin.Plugin.Link(self)
        
        numTextures = len(self.iTextureList)
        for i in range(numTextures):
            texture = self.iTextureList[i]
            self.iTextureList[i] = TextureManager.GetTextureManager().TextureByNameOrLoad(texture)
        self.iLinked = True
               
    def Vector3d(self, aVector3d):
        self.iTextureIndex = aVector3d.iX
        if self.iTextureIndex > len(self.iTextureList) - 1:
            self.iTextureIndex = len(self.iTextureList) - 1
        if self.iTextureIndex < 0:
            self.iTextureIndex = 0
        self.iPlugin.Texture(self.iTextureList[self.iTextureIndex])
        
    def TextureAtIndex(self, aTextureIndex):
        return self.iTextureList[aTextureIndex]
    
    def SetTexture(self, aTexture, aTextureIndex=0):
        self.iTextureList[aTextureIndex] = aTexture
        if aTextureIndex == self.iTextureIndex:
            self.iPlugin.Texture(aTexture)
        
    def NumTextures(self):
        return len(self.iTextureList)
    
    def SetNumTextures(self, aNumTextures):
        if aNumTextures < self.NumTextures():
            self.iTextureList = self.iTextureList[0:aNumTextures]
        else:
            numToAdd = aNumTextures - self.NumTextures()
            for i in range(numToAdd):
                self.iTextureList.append('')
        if self.iTextureIndex > aNumTextures:
            self.iTextureIndex = aNumTextures - 1
            self.iPlugin.Texture(self.iTextureList[self.iTextureIndex])
        self.ObserverUpdate()

        
class TextureArrayFixed(TextureArray):
    def __init__(self, aName=None, aTextureList=None, aFps=None, aLoop=None):
        TextureArray.__init__(self, aName, aTextureList, aFps, aLoop)
        
    def Save(self, aStream):
        if aStream.ContextAttribute('class') == None:
            aStream.SetContextAttribute('class', 'TextureArrayFixed')
            
        TextureArray.Save(self, aStream)
        
        
        
        

File: Public/Gui/Linn/Gui/Resources/TextureArray.cs
using Linn.Gui.Scenegraph;
using Linn.Gui.Resources;
using System.Collections.Generic;
using System.Xml;

namespace Linn {
namespace Gui {
    
public class TextureArrayFixed : Plugin
{
    public TextureArrayFixed() : base("TextureArrayFixed") {
    }
    
    public TextureArrayFixed(string aName) : base(aName) {
    }
    
    public override void Load(XmlNode aXmlNode) {
        base.Load(aXmlNode);
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Texture");
        if(list != null) {
            foreach(XmlNode node in list) {
                ReferenceTexture texture = new ReferenceTexture(node.FirstChild.Value);
                iTextureList.Add(texture);
            }
        }
    }
    
    public override void Link() {
        base.Link();
        foreach(ReferenceTexture texture in iTextureList) {
            texture.Link();
        }
        // NOTE: should texture array class store index state???
        if(NextPlugin != null) {
            if(iTextureList.Count > 0) {
                NextPlugin.Texture(iTextureList[iTextureIndex]);
            } else {
                NextPlugin.Texture(new ReferenceTexture());
            }
        }
    }
    
    public override void Save(XmlTextWriter aWriter) {
        base.Save(aWriter);
        foreach(ReferenceTexture t in iTextureList) {
            aWriter.WriteStartElement("Texture");
            aWriter.WriteString(t.Object.Filename);
            aWriter.WriteEndElement();
        }
    }
    
    public override void Vector3d(Vector3d aVector) {
        iTextureIndex = (int)aVector.X;
        if(iTextureIndex < 0){
            iTextureIndex = 0;
        } else if(iTextureIndex > iTextureList.Count - 1) {
            iTextureIndex = iTextureList.Count - 1;
        }
        if(NextPlugin != null) {
            if(iTextureList.Count > 0) {
                NextPlugin.Texture(iTextureList[iTextureIndex]);
            } else {
                NextPlugin.Texture(new ReferenceTexture());
            }
        }
    }
    
    public override bool ProcessMessage(Message aMessage) {
        if(aMessage.Fullname == Fullname) {
            MsgSetTexture setTexture = aMessage as MsgSetTexture;
            if(setTexture != null) {
                this[setTexture.Index] = setTexture.Texture;
            }
            return true;
        }
        if(base.ProcessMessage(aMessage)) {
            return true;
        }
        return false;
    }
    
    public ITexture this[int aIndex] {
        get {
            if(iTextureList.Count > 0) {
                return ((ReferenceTexture)iTextureList[aIndex]).Object;
            } else {
                return null;
            }
        }
        set {
            if(aIndex >= 0 && aIndex < iTextureList.Count)
            {
                ReferenceTexture texture = new ReferenceTexture(value);
                iTextureList[aIndex] = texture;
                if(aIndex == iTextureIndex) {
                    if(iTextureList.Count > 0) {
                        NextPlugin.Texture(iTextureList[iTextureIndex]);
                    } else {
                        NextPlugin.Texture(new ReferenceTexture());
                    }
                }
                ObserverUpdate();
            }
        }
    }
    
    public List<ReferenceTexture> Textures {
        get {
            return iTextureList;
        }
    }
    
    public void AddTexture(ITexture aTexture) {
        iTextureList.Add(new ReferenceTexture(aTexture));
        if(iTextureList.Count - 1 == iTextureIndex) {
            if(NextPlugin != null) {
                if(iTextureList.Count > 0) {
                    NextPlugin.Texture(iTextureList[iTextureIndex]);
                } else {
                    NextPlugin.Texture(new ReferenceTexture());
                }
            }
        }
        ObserverUpdate();
    }
    
    public void RemoveTexture(ITexture aTexture) {
        ReferenceTexture obj = null;
        foreach(ReferenceTexture t in iTextureList) {
            if(t.Object == aTexture) {
                obj = t;
                break;
            }
        }
        bool updateTexture = false;
        if(aTexture == this[iTextureIndex]) {
            updateTexture = true;
        }
        iTextureList.Remove(obj);
        if(iTextureList.Count < iTextureIndex) {
            iTextureIndex = iTextureList.Count - 1;
            if(NextPlugin != null) {
                if(iTextureList.Count > 0) {
                    NextPlugin.Texture(iTextureList[iTextureIndex]);
                } else {
                    NextPlugin.Texture(new ReferenceTexture());
                }
            }
        } else if(updateTexture) {
            if(NextPlugin != null) {
                if(iTextureList.Count > 0) {
                    NextPlugin.Texture(iTextureList[iTextureIndex]);
                } else {
                    NextPlugin.Texture(new ReferenceTexture());
                }
            }
        }
        ObserverUpdate();
    }
    
    public void Clear() {
        iTextureList.Clear();
        iTextureIndex = 0;
        NextPlugin.Texture(new ReferenceTexture());
        ObserverUpdate();
    }
    
    protected List<ReferenceTexture> iTextureList = new List<ReferenceTexture>();
    protected int iTextureIndex = 0;
}

public class TextureArray : TextureArrayFixed
{
    public TextureArray() : base("TextureArray") {
    }
    
    public TextureArray(string aName) : base(aName) {
    }
    
    public void SetNumTextures(int aNumTextures) {
        if(aNumTextures < iTextureList.Count) {
            iTextureList.RemoveRange(aNumTextures, iTextureList.Count - 1);
        } else {
            int numToAdd = aNumTextures - iTextureList.Count;
            for(int i = 0; i < numToAdd; ++i) {
                iTextureList.Add(new ReferenceTexture());
            }
        }
        if(iTextureIndex > aNumTextures) {
            iTextureIndex = iTextureList.Count - 1;
            if(iTextureList.Count > 0) {
                NextPlugin.Texture(iTextureList[iTextureIndex]);
            } else {
                NextPlugin.Texture(new ReferenceTexture());
            }
            ObserverUpdate();
        }
    }
}
    
} // Gui
} // Linn
